### DISCUSSION TITLE
Moving files

### DISCUSSION THREAD

**Discussion by aneshodza** [nvim-tree/nvim-tree.lua#2292] [Q&A]
I saw that there is no "move" function on nvim-tree. Was that an oversight from me, or is there really no move?

**Can this functionality be implemented utilising API?**
It should be possible, as we can already cut and paste.

**Is your feature request related to a problem? Please describe.**
I wanted to move a file from `something/else/file.lua` to `something/file.lua`. You can achieve that by cutting and pasting, but why not have both in one step.

**Describe the solution you'd like**
Just a move function. The window for moving it to somewhere else could look the same as the rename but also show the path that can be altered.

**Comment by joelpeapen** [comment]
It's on the wiki: https://github.com/nvim-tree/nvim-tree.lua/wiki/Recipes#nerdtree-style-copy-file-to

```lua
local function on_attach(bufnr)
...
    local function move_file_to()
        local node = api.tree.get_node_under_cursor()
        local file_src = node['absolute_path']
        local file_out = vim.fn.input("MOVE TO: ", file_src, "file")
        local dir = vim.fn.fnamemodify(file_out, ":h")
        vim.fn.system { 'mkdir', '-p', dir }
        vim.fn.system { 'mv',  file_src, file_out }
    end
    vim.keymap.set('n', 'X', move_file_to, opts('Move File To'))
end
```

**Comment by aneshodza** [comment]
But this is still a copy (it doesn't remove the old location). This function does a `cp`, instead of an `mv`
Now:
```lua
vim.fn.system { 'cp', '-R', file_src, file_out }
```
To move:
```lua
vim.fn.system { 'mv', '-R', file_src, file_out }
```
But the message I'm getting from this, is that the lack of a move is a design choice?

**Comment by joelpeapen** [comment]
removing the  ```'-R'``` flag works

**Comment by aneshodza** [comment]
But `cp` is still a copy. I'm talking about having a built in `move_file_to`. Maybe by just copying and slightly changing `copy_file_to`.

**Comment by joelpeapen** [comment]
yes you can rename the function. I've updated the snippet.

**Comment by aneshodza** [comment]
Alright, thanks! Im wondering why this isn't in per default.

**Comment by alex-courtis** [comment]
Many thanks @joelpeapen

**Comment by svanharmelen** [comment]
One more question about this solution... If I rename a file I get a nice "in place" popup where I can type the name. Is it possible to use such a popup for this custom function as well @joelpeapen?

**Reply by cowboysmall** [reply]
**EDIT:** this approach is incorrect - please see later implementation.

Not sure if you are still inetrested, or if you figured this out for yourself, but I wanted the same thing - here is my solution:

```lua
local map = vim.keymap.set

local nvim_tree = require("nvim-tree.api")

local function copy_file_to()
  local file_src = nvim_tree.tree.get_node_under_cursor()["absolute_path"]
  local input_opts = { prompt = "Copy to ", default = file_src, completion = "file" }

  vim.ui.input(input_opts, function(file_out)
    local dir = vim.fn.fnamemodify(file_out, ":h")

    vim.fn.system { "mkdir", "-p", dir }
    vim.fn.system { "cp", "-R", file_src, file_out }
  end)
end

local function move_file_to()
  local file_src = nvim_tree.tree.get_node_under_cursor()["absolute_path"]
  local input_opts = { prompt = "Move to ", default = file_src, completion = "file" }

  vim.ui.input(input_opts, function(file_out)
    local dir = vim.fn.fnamemodify(file_out, ":h")

    vim.fn.system { "mkdir", "-p", dir }
    vim.fn.system { "mv", file_src, file_out }
  end)
end

map("n", "<M-c>", copy_file_to, { desc = "copy file to", remap = true })
map("n", "<M-x>", move_file_to, { desc = "move file to", remap = true })
```

I placed this code in the `mappings.lua` within `.config/nvim/lua` (I'm using nvchad). 

I'm quite new to both neovim and lua, so there may be a better way to do this, but it works for me exactly as I wanted it.

Hope that helps.

**EDIT:** this approach is incorrect - please see later implementation.

**Reply by svanharmelen** [reply]
Thanks @cowboysmall, much appreciated!

**Reply by cowboysmall** [reply]
Actually, it appears I have indeed found a better way to do this - after reading more documentation, and looking more closely at the recipes and plugin code, I have settled on this approach:

```lua
      on_attach = function(bufnr)
        local api = require("nvim-tree.api")

        -- default mappings
        api.config.mappings.default_on_attach(bufnr)

        local function opts(desc)
          return { desc = "nvim-tree: " .. desc, buffer = bufnr, noremap = true, silent = true, nowait = true }
        end

        local function copy_file_to()
          local file_src = api.tree.get_node_under_cursor()["absolute_path"]
          local input_opts = { prompt = "Copy to ", default = file_src, completion = "file" }

          vim.ui.input(input_opts, function(file_out)
            local dir = vim.fn.fnamemodify(file_out, ":h")

            vim.fn.system { "mkdir", "-p", dir }
            vim.fn.system { "cp", "-R", file_src, file_out }
          end)
        end

        local function move_file_to()
          local file_src = api.tree.get_node_under_cursor()["absolute_path"]
          local input_opts = { prompt = "Move to ", default = file_src, completion = "file" }

          vim.ui.input(input_opts, function(file_out)
            local dir = vim.fn.fnamemodify(file_out, ":h")

            vim.fn.system { "mkdir", "-p", dir }
            vim.fn.system { "mv", file_src, file_out }
          end)
        end

        -- custom mappings
        local map = vim.keymap.set

        map("n", "<M-c>", copy_file_to, opts("copy file to"))
        map("n", "<M-x>", move_file_to, opts("move file to"))
      end
```

this code snippet is located in the **nvim-tree** plugin config:

```lua
return {

  {
    "nvim-tree/nvim-tree.lua",

    opts = {

      view = {
        width = 50
      },

      on_attach = function(bufnr)
        ...
      end
    }
  }

}
```

so this is now much more consistent with the suggested approach, and makes use of provided examples of how to override mappings, etc.

I wasn't kidding when I said I was new to neovim and lua ;-)

**Reply by alex-courtis** [reply]
That's fantastic, I'd really appreciate your adding this to recipes - please attribute yourself as the author.

Little improvement:
You can check the system call. If there was a problem you can output an error and halt.
```lua
      local res = vim.fn.system({ "mkdir", "-p", dir })
      if vim.v.shell_error ~= 0 then
        vim.notify(res, vim.log.levels.ERROR, { title = "NvimTree" })
        return
      end
```

**Reply by cowboysmall** [reply]
That's a great improvement - I'll remember that for any future enhancements that I make. I definitely have much to learn about neovim and lua, but I'm enjoying my journey so far.

```lua
        map("n", "<M-x>", function()
          local file_src = api.tree.get_node_under_cursor()["absolute_path"]
          local input_opts = { prompt = "Move to ", default = file_src, completion = "file" }

          vim.ui.input(input_opts, function(file_out)
            local dir = vim.fn.fnamemodify(file_out, ":h")

            local res = vim.fn.system({ "mkdir", "-p", dir })
            if vim.v.shell_error ~= 0 then
              vim.notify(res, vim.log.levels.ERROR, { title = "NvimTree" })
              return
            end

            vim.fn.system { "mv", file_src, file_out }
          end)
        end, opts("Move File To"))
```

Should I addit to the end of the recipes page? Or should I add it close to the other relevent "NERDTree Style Copy File To" entry?

**Reply by alex-courtis** [reply]
I reckon move it before that one.

I'll go through the page and move superseded or now-core-functionality ones to a new section at the end.

**Reply by cowboysmall** [reply]
Added. I included a link to the relevant file in my dotfiles.

Thanks again.
